packages/flutter/lib/src/widgets/_web_image_web.dart

```diff
@@ -0,0 +1,241 @@
+    });
+  }
+
+  HtmlElementView? htmlElementView;
+
+  num? naturalWidth;
+  num? naturalHeight;
+
+  int? imgElementId;
+
+  @override
+  void initState() {
+    super.initState();
+    _cachedImage =
+        _WebImageCache._getOrCreateCachedImage(widget.src);
+    _cachedImage.state.addListener(_onCachedImageStateChange);
+  }
+
+  @override
+  void dispose() {
+    super.dispose();
+    _cachedImage.state.removeListener(_onCachedImageStateChange);
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    final _CachedImageInfo info =
+        _WebImageCache._getOrCreateCachedImage(widget.src);
+    final double naturalWidth = info.image.naturalWidth.toDouble();
+    final double naturalHeight = info.image.naturalHeight.toDouble();
+    return ValueListenableBuilder<_ImageLoadingState>(
+      valueListenable: info.state,
+      builder: (BuildContext context, _ImageLoadingState state, Widget? child) {
+        final Widget? builtWidget = switch (state) {
+          _ImageLoadingState.loading =>
+            widget.loadingBuilder?.call(context, child!, null),
+          _ImageLoadingState.success =>
+            widget.frameBuilder?.call(context, child!, 0, true),
+          _ImageLoadingState.error =>
+            widget.errorBuilder?.call(context, info.error, null) ?? Container(),
+        };
+        return builtWidget ?? child!;
+      },
+      child: FittedBox(
+        fit: BoxFit.scaleDown,
+        child: SizedBox(
+          width: naturalWidth,
+          height: naturalHeight,
+          child: _buildHtmlImage(),
+        ),
+      ),
+    );
+  }
+
+  Widget _buildHtmlImage() {
+    return HtmlElementView(
+      viewType: _viewType,
+      creationParams: <String, String>{'src': widget.src},
+      hitTestBehavior: PlatformViewHitTestBehavior.transparent,
+    );
+  }
+}
+
+class _WebImageCache {
+  static final Map<String, _CachedImageInfo> _imageCache =
+      <String, _CachedImageInfo>{};
+
+  static _CachedImageInfo _getOrCreateCachedImage(String url) {
+    _CachedImageInfo? info = _imageCache[url];
+    if (info == null) {
+      info = _CachedImageInfo.fromUrl(url);
+      _imageCache[url] = info;
+    }
+    return info;
+  }
+}
+
+web.HTMLImageElement _createImage(String url) {
+  final web.HTMLImageElement image =
+      web.document.createElement('img') as web.HTMLImageElement;
+  image.src = url;
+  image.style
+    ..width = '100%'
+    ..height = '100%';
+  return image;
+}
+
+enum _ImageLoadingState {
+  loading,
+  success,
+  error,
+}
+
+class _CachedImageInfo {
+  _CachedImageInfo.fromUrl(String url) : image = _createImage(url) {
+    _startDecoding();
+  }
+
+  final web.HTMLImageElement image;
+
+  ValueNotifier<_ImageLoadingState> state =
+      ValueNotifier<_ImageLoadingState>(_ImageLoadingState.loading);
+
+  Object? _error;
+  Object get error {
+    assert(hasError);
+    return _error!;
+  }
+
+  bool get isLoading => state.value == _ImageLoadingState.loading;
```

Comment:
The getters isLoading and isSuccess aren't used? Also, the hasError getter _could_ be inlined?

Reply:
Done.